sst/SimengCoreWrapper.cc

```diff
@@ -0,0 +1,218 @@
+      mode = SimulationMode::InOrderPipelined;
+      modeString = "Emulation";
+    } else if (config["Core"]["Simulation-Mode"].as<std::string>() ==
+               "outoforder") {
+      mode = SimulationMode::OutOfOrder;
+      modeString = "Out-of-Order";
+    } else {
+      mode = SimulationMode::Emulation;
+      modeString = "Emulation";
+    }
+
+    float clockFreq_ = config["Core"]["Clock-Frequency"].as<float>();
+    uint32_t timerFreq_ = config["Core"]["Timer-Frequency"].as<uint32_t>();
+    timer_modulo = (clockFreq_ * 1e9) / (timerFreq_ * 1e6);
+
+    // // Create the process Image
+    std::vector<std::string> commandLine({executable_path, executable_args});
+    process = std::make_unique<simeng::kernel::LinuxProcess>(commandLine, config);
+    if (!process->isValid()) output.fatal(CALL_INFO, 1, 0, "Could not read/parse %s", executable_path);
+
+    auto processImage = process->getProcessImage();
+    size_t processMemorySize = processImage.size();
+    process_memory = new char[processMemorySize]();
+    std::copy(processImage.begin(), processImage.end(), process_memory);
+
+    uint64_t entryPoint = process->getEntryPoint();
+
+    // Create the OS kernel with the process
+    kernel = std::make_unique<simeng::kernel::Linux>();
+    kernel->createProcess(*process.get());
+
+    instruction_memory = std::make_unique<simeng::FlatMemoryInterface>(
+        process_memory, processMemorySize);
+
+    // // Create the architecture, with knowledge of the kernel
+    arch =
+        std::make_unique<simeng::arch::aarch64::Architecture>(*kernel, config);
+    
+    predictor = std::make_unique<simeng::BTBPredictor>(
+        config["Branch-Predictor"]["BTB-bitlength"].as<uint8_t>());
+    auto config_ports = config["Ports"];
+    std::vector<std::vector<uint16_t>> portArrangement(config_ports.size());
+    // Extract number of ports
+    for (size_t i = 0; i < config_ports.size(); i++) {
+      auto config_groups = config_ports[i]["Instruction-Group-Support"];
+      // Extract number of groups in port
+      for (size_t j = 0; j < config_groups.size(); j++) {
+        portArrangement[i].push_back(config_groups[j].as<uint16_t>());
+      }
+    }
+    port_allocator = std::make_unique<simeng::pipeline::BalancedPortAllocator>(
+        portArrangement);
+
+    // Configure reservation station arrangment
+    std::vector<std::pair<uint8_t, uint64_t>> rsArrangement;
+    for (size_t i = 0; i < config["Reservation-Stations"].size(); i++) {
+      auto reservation_station = config["Reservation-Stations"][i];
+      for (size_t j = 0; j < reservation_station["Ports"].size(); j++) {
+        uint8_t port = reservation_station["Ports"][j].as<uint8_t>();
+        if (rsArrangement.size() < port + 1) {
+          rsArrangement.resize(port + 1);
+        }
+        rsArrangement[port] = {i, reservation_station["Size"].as<uint16_t>()};
+      }
+    }
+
+    modeString = "Out-of-Order";
```

Comment:
PTAL, I've added support for all modes.

Reply:
We can support emulation mode and out-of-order mode, but not inorder mode because it requires a FlatMemoryInterface& as a constructor argument. Should inorder core constructor be changed to accept MemoryInterface?

Reply:
The inorder core needs an update but I think that may come with Dan's RISCV work so leave for now. However, if this PR will only support SST+OoO core then we'll need to prevent the simulation from starting if the passed config file doesn't use the OoO core.